mycpp/control_flow_pass.py

```diff
@@ -0,0 +1,418 @@
+"""
+control_flow_pass.py - AST pass that builds a control flow graph.
+"""
+from typing import overload, Union, Optional, Dict
+
+import mypy
+from mypy.visitor import ExpressionVisitor, StatementVisitor
+from mypy.nodes import (Block, Expression, Statement, ExpressionStmt, StrExpr,
+                        ForStmt, WhileStmt, CallExpr, FuncDef, IfStmt)
+
+from mypy.types import Type
+
+from mycpp.crash import catch_errors
+from mycpp.util import split_py_name
+from mycpp import util
+from mycpp import pass_state
+
+T = None  # TODO: Make it type check?
+
+
+class UnsupportedException(Exception):
+    pass
+
+
+class Build(ExpressionVisitor[T], StatementVisitor[None]):
+
+    def __init__(self, types: Dict[Expression, Type]):
+
+        self.types = types
+        self.cfgs = {}
+        self.current_statement_id = None
+        self.current_func_node = None
+        self.loop_stack = []
+
+    def current_cfg(self):
+        if not self.current_func_node:
+            return None
+
+        return self.cfgs.get(split_py_name(self.current_func_node.fullname))
+
+    #
+    # COPIED from IRBuilder
+    #
+
+    @overload
+    def accept(self, node: Expression) -> T:
+        ...
+
+    @overload
+    def accept(self, node: Statement) -> None:
+        ...
+
+    def accept(self, node: Union[Statement, Expression]) -> Optional[T]:
+        with catch_errors(self.module_path, node.line):
+            if isinstance(node, Expression):
+                try:
+                    res = node.accept(self)
+                    #res = self.coerce(res, self.node_type(node), node.line)
+
+                # If we hit an error during compilation, we want to
+                # keep trying, so we can produce more error
+                # messages. Generate a temp of the right type to keep
+                # from causing more downstream trouble.
+                except UnsupportedException:
+                    res = self.alloc_temp(self.node_type(node))
+                return res
+            else:
+                try:
+                    cfg = self.current_cfg()
+                    if cfg and not isinstance(node, Block) and not isinstance(node, ForStmt) and not isinstance(node, WhileStmt):
```

Comment:
OK I didn't notice this in the previous incarnation -- the accept() has some logic

That is interesting, maybe put a comment here to explain it